### PR TITLE
Specific

### PR DESCRIPTION
**Pull request by lee-kagiso**
- Return the specific page rather than generic Page
- Removed surplus code that a) doesn't work as it assumes a correct spelling of a slug (as it doesn't use trigrams and in fact does a look up by slug), which by definition doesn't work when the slug is misspelled! Also, the existing methods are adequate to do an instant redirect, as if only 1 page is suggested, auto redirect to that one.

### CODE REVIEW

File: kagiso_smart_404/views.py [hunk 1]
```diff
@@ -11,11 +11,11 @@ def not_found(request):  # pragma: no cover
     slug = request.path
     root_page = request.site.root_page.specific
 
-    exact_match = slug_matches_one_page_exactly(slug, root_page)
-    if exact_match:
-        return HttpResponsePermanentRedirect(redirect_to=exact_match.url)
-
     suggested_pages = suggest_page_from_misspelled_slug(slug, root_page)
 
+    if suggested_pages and len(suggested_pages) == 1:
```
Comment (lee-kagiso) [inline kagiso_smart_404/views.py]: This replaces all the removed code...

File: kagiso_smart_404/views.py [hunk 2]
```diff
@@ -2,7 +2,7 @@
 from django.shortcuts import render
 
 from .utils import (
-    slug_matches_one_page_exactly,
+    # slug_matches_one_page_exactly,
```
Comment (colinfrankb) [inline kagiso_smart_404/views.py]: This function was deleted.

**Comment by colinfrankb** [comment]
I understand the definition of a trigram , but can you elaborate on this : "as it doesn't use trigrams" and how it is relevant

File: kagiso_smart_404/utils.py [hunk 3]
```diff
@@ -1,49 +1,24 @@
...
-                AND first_published_at IS NOT NULL
-                AND substring(path FOR %(root_path_len)s) = %(root_path)s
-                AND slug %% %(slug)s
-             ORDER BY similarity DESC
-             '''
+    FROM wagtailcore_page p
+    WHERE live = true
+    AND first_published_at IS NOT NULL
+    AND substring(path FOR %(root_path_len)s) = %(root_path)s
+    AND slug %% %(slug)s
```
Comment (lee-kagiso) [inline kagiso_smart_404/utils.py:13]: this is pg trigram syntax

**Comment by lee-kagiso** [comment]
if the slug is mispelled, you have to use some technique to find similar slugs and trigrams do that. Otherwise all you have is a mispelled slug and you can do nothing with it. So the previous code only worked if the slug was correct, which is obviously not the reality of a 404

File: kagiso_smart_404/utils.py (hunk 3 above)
Comment (colinfrankb) [inline kagiso_smart_404/utils.py:13]: cool

File: kagiso_smart_404/utils.py [hunk 4]
```diff
@@ -1,49 +1,24 @@
...
     data = {
         'slug': slug,
         'root_path': root_path,
         'root_path_len': root_path_len
     }
-    suggested_pages = Page.objects.raw(sql, data)
+    suggested_pages = list(Page.objects.raw(sql, data))  # RawQuerySet...
+    suggested_pages = [page.specific for page in suggested_pages]
 
     return suggested_pages[:3] if suggested_pages else None
```
Comment (colinfrankb) [inline kagiso_smart_404/utils.py]: just use `return suggested_pages[:3]`

File: kagiso_smart_404/views.py (hunk 1 above)
Comment (colinfrankb) [inline kagiso_smart_404/views.py]: then you can just use `if len(suggested_pages) == 1`

File: kagiso_smart_404/utils.py (hunk 4 above)
Comment (lee-kagiso) [inline kagiso_smart_404/utils.py]: yeah you right, my new code always returns a list

**Pull request merged**